libOTe_Tests/SilentOT_Tests.cpp

```diff
@@ -928,6 +929,220 @@ void Tools_Pprf_test(const CLP& cmd)
 #endif
 }
 
+// union conv128 {
+//     unsigned __int128 u;
+//     block m;
+// };
+
+// struct alignas(16) u128 {
+//     unsigned __int128 value;
+//     // Constructors
+//     u128() : value(0) {}
+//     u128(unsigned __int128 val) : value(val) {}
+//     u128(block b) {
+//         conv128 c;
+//         c.m = b;
+//         value = c.u;
+//     }
+
+//     // Conversion operator to unsigned __int128
+//     operator unsigned __int128() const {
+//         return value;
+//     }
+
+//     // Additional member function
+//     void customFunction() {
+//         std::cout << "This is a custom function for u128!" << std::endl;
+//     }
+// };
+
+using u128 = unsigned __int128;
```

Comment:
not portable.

Reply:
The reason I use this is if I use the class version commented above, I would get error

```
In file included from /home/halulu/dev/libOTe/out/install/linux/include/coproto/Socket/Socket.h:20,
                 from /home/halulu/dev/libOTe/out/install/linux/include/coproto/coproto.h:11,
                 from /home/halulu/dev/libOTe/libOTe_Tests/../libOTe/Tools/Coproto.h:9,
                 from /home/halulu/dev/libOTe/libOTe_Tests/../libOTe/Tools/SilentPprf.h:20,
                 from /home/halulu/dev/libOTe/libOTe_Tests/SilentOT_Tests.cpp:3:
/home/halulu/dev/libOTe/out/install/linux/include/coproto/Proto/Buffers.h: In instantiation of ‘macoro::coroutine_handle<void> coproto::internal::MoveSendProto<Container>::await_suspend(const H&) [with H = macoro::coroutine_handle<macoro::detail::task_promise<void, true> >; Container = std::vector<std::array<u128, 4>, std::allocator<std::array<u128, 4> > >]’:
/home/halulu/dev/libOTe/out/install/linux/include/macoro/coro_frame.h:152:29:   required from ‘auto macoro::await_suspend(Awaiter&, coroutine_handle<T>, enable_if_t<((! has_void_await_suspend<Awaiter, coroutine_handle<T> >::value) && (! has_bool_
await_suspend<Awaiter, coroutine_handle<T> >::value)), empty_state>) [with Awaiter = coproto::internal::MoveSendProto<std::vector<std::array<u128, 4>, std::allocator<std::array<u128, 4> > > >; T = detail::task_promise<void, true>; enable_if_t<((! has_void_await_suspend<Awaiter, coroutine_handle<T> >::value) && (! has_bool_await_suspend<Awaiter, coroutine_handle<T> >::value)), empty_state> = empty_state]’
/home/halulu/dev/libOTe/libOTe_Tests/../libOTe/Tools/SubfieldPprf.h:621:25:   required from ‘macoro::task<void> osuCrypto::SilentSubfieldPprfSender<G, F>::Expander::run() [with G = u128; F = u128]’
/home/halulu/dev/libOTe/libOTe_Tests/../libOTe/Tools/SubfieldPprf.h:300:75:   required from ‘macoro::task<void> osuCrypto::SilentSubfieldPprfSender<G, F>::expand(osuCrypto::Socket&, osuCrypto::span<const T>, osuCrypto::PRNG&, osuCrypto::MatrixView<F>, osuCrypto::PprfOutputFormat, bool, osuCrypto::u64) [with G = u128; F = u128; osuCrypto::Socket = coproto::Socket; osuCrypto::span<const T> = nonstd::span_lite::span<const u128, 18446744073709551615>; osuCrypto::u64 = long unsigned int]’    
/home/halulu/dev/libOTe/libOTe_Tests/SilentOT_Tests.cpp:1104:28:   required from here
/home/halulu/dev/libOTe/out/install/linux/include/coproto/Proto/Buffers.h:260:53: error: no matching function for call to ‘asSpan(std::vector<std::array<u128, 4>, std::allocator<std::array<u128, 4> > >&)’
  260 |                                 if (internal::asSpan(mContainer).size() == 0)
      |                                     ~~~~~~~~~~~~~~~~^~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
```

It seems chl cannot send it directly, how should I fix this?

Reply:
so if you send data, then the thing `T` you're sending either needs to be `std::is_trivial<T>::value == true` or you can send something that looks like a vector of trivials things, e.g.  `std::vector<T>` where `T` is trivial. Specifically, when you pass something to coproto, it checks this using https://github.com/Visa-Research/coproto/blob/main/coproto/Common/TypeTraits.h#L160 

If this trait is true for your type, then that function `internal::asSpan(mContainer)` exists here https://github.com/Visa-Research/coproto/blob/main/coproto/Proto/Operation.h#L122

Maybe I should make this error message a bit more descriptive ;)